### PR TITLE
Fold functional component trees that have a class component as root

### PR DESCRIPTION
**Pull request by trueadm**
Currently, you can only fold functional components. This PR aims to allow functional component folding directly into a class component. Essentially, this is an incremental step to adding full class component folding.

### CODE REVIEW

File: src/react/components.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import * as t from "babel-types";
+import type { BabelNodeIdentifier } from "babel-types";
+import { createAbstractObject } from "../flow/abstractObjectFactories.js";
+import { valueIsClassComponent } from "./utils";
+import { ExpectedBailOut } from "./reconcilation.js";
+import { Construct, Get, GetValue } from "../methods/index.js";
+import { Properties } from "../singletons.js";
+import invariant from "../invariant.js";
+
+export function getInitialProps(
```
Comment (trueadm) [inline src/react/components.js]: This was moved from `reconciliation`.

File: src/react/components.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+          throw new ExpectedBailOut(`root component missing Flow type annotations for the "props" paramater`);
+        }
+        propTypes = flowAnnotationToObjectTypeTemplate(propsTypeAnnotation);
+      }
+    }
+  }
+  return createAbstractObject(realm, propsName, propTypes);
+}
+
+export function getInitialContext(
```
Comment (trueadm) [inline src/react/components.js]: This was moved from `reconciliation`.

File: src/flow/abstractObjectFactories.js [hunk 3]
```diff
@@ -71,14 +71,14 @@ function _createAbstractObject(
   template.makeSimple();
   value.values = new ValuesDomain(new Set([template]));
   realm.rebuildNestedProperties(value, name);
-  return value;
+  return ((value: any): AbstractObjectValue);
```
Comment (NTillmann) [inline src/flow/abstractObjectFactories.js]: Adding an `invariant(value instanceof AbstractObjectValue);` would be nice.

File: src/react/components.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import invariant from "../invariant.js";
+
+export function getInitialProps(
+  realm: Realm,
+  componentType: ECMAScriptSourceFunctionValue
+): ObjectValue | AbstractObjectValue {
+  let propsName = null;
+  let propTypes = null;
+  if (valueIsClassComponent(realm, componentType)) {
+    propsName = "this.props";
```
Comment (NTillmann) [inline src/react/components.js]: I wonder if `this...` is going to be always bound to the right thing in the code generation.

File: src/react/components.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+export function createClassInstance(
+  realm: Realm,
+  componentType: ECMAScriptSourceFunctionValue,
+  props: ObjectValue | AbstractValue,
+  context: ObjectValue | AbstractValue
+): AbstractObjectValue {
+  let componentPrototype = Get(realm, componentType, "prototype");
+  invariant(componentPrototype instanceof ObjectValue);
+  let instance = createAbstractObject(realm, "this", null);
```
Comment (NTillmann) [inline src/react/components.js]: I wonder if `this` is always going to get bound to the right thing.

File: src/react/components.js (hunk 4 above)
Comment (trueadm) [reply src/react/components.js]: It seems to work in all the tests.

**Pull request closed without merging**